### PR TITLE
Add batched and strided batched gemm as TunableOp

### PR DESCRIPTION
**Pull request by cloudhan**

### CODE REVIEW

File: onnxruntime/python/tools/kernel_explorer/device_array.h [hunk 1]
```diff
@@ -37,28 +38,26 @@ class DeviceArray {
...
-  ~DeviceArray() {
-    CALL_THROW(FREE(x_device_));
+    return device_.get();
   }
 
  private:
-  void* x_device_;
-  void* x_host_;
+  std::shared_ptr<void> device_;
+  void* host_;
```
Comment (abudup) [inline onnxruntime/python/tools/kernel_explorer/device_array.h:60]: Can you please elaborate why `device_` is a `shared_ptr` and `host_` is not?

File: onnxruntime/core/providers/rocm/tunable/gemm.cu [hunk 2]
```diff
@@ -70,13 +69,141 @@ inline GEMM(T, ScalarT) {
...
+  }
+  params.as = as;
+  params.lda = lda;
+  params.bs = bs;
+  params.ldb = ldb;
+  if constexpr (!std::is_same_v<T, ScalarT> && std::is_same_v<ScalarT, float>) {
+    params.beta = ToHipType<T>::FromFloat(std::forward<T>(beta));
+  } else {
+    params.beta = beta;
+  }
```
[the remark this answers is not in the document]
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/tunable/gemm.cu:96]: This actually compiles, and it is in our code now and works very well. Why do you think it wont compile?

File: onnxruntime/python/tools/kernel_explorer/device_array.h (hunk 1 above)
Comment (cloudhan) [reply onnxruntime/python/tools/kernel_explorer/device_array.h:60]: DeviceArray here is actually a DeviceArrayAndHostArrayView. And the device part just (manually) mirrors the host part.

File: onnxruntime/core/providers/rocm/tunable/gemm.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (abudup) [reply onnxruntime/core/providers/rocm/tunable/gemm.cu:96]: Heh heh, I was pretty sure I missed a `>` or a `typename` somewhere :-) But glad that it works!

File: onnxruntime/contrib_ops/cuda/bert/decoder_attention.cc [hunk 3]
```diff
@@ -363,6 +363,9 @@ Status DecoderAttention<T>::ComputeInternal(OpKernelContext* context) const {
 
   return LaunchDecoderAttentionKernel(
           device_prop,
+#ifdef USE_ROCM
```
[the remark this answers is not in the document]
Comment (cloudhan) [reply onnxruntime/contrib_ops/cuda/bert/decoder_attention.cc:366]: We don't comment these type of thing. There are too many of them are scattering in the CUDA EP make it impractically to do so. Basically, when you see something like this one, they just are useful after hipify. Otherwise, useless for CUDA.

File: onnxruntime/core/providers/rocm/tunable/gemm.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/tunable/gemm.cu:96]: Since this is `if constexpr`, we don't need thoes template tricks anymore.

File: onnxruntime/core/providers/rocm/tunable/gemm_tunable.cuh [hunk 4]
```diff
@@ -69,6 +68,71 @@ class GemmTunableOp : public tunable::TunableOp<GemmParams<T>> {
...
+
+  const BatchedGemmParams<T>* PreTuning(const BatchedGemmParams<T>* params) override {
+    if (!IsZero(params->beta)) {
+      // See GemmTunableOp<T>::PreTuning for more details
+      BatchedGemmParams<T>* proxy = new BatchedGemmParams<T>();
+      *proxy = *params;
+      using PointerOfT = T*;
+      proxy->cs = new PointerOfT[params->batch];
+      for (int i = 0; i < params->batch; i++) {
+        HIP_CALL_THROW(hipMalloc(&(proxy->cs[i]), proxy->m * proxy->ldc * sizeof(T)));
```
[the remark this answers is not in the document]
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/tunable/gemm_tunable.cuh:86]: Good catch, BatchedGemm::PreTuning is current bugged. The as, bs, cs are device buffers of device pointers. I took them as host buffers of device pointers. This also needs a fix.

File: onnxruntime/core/providers/rocm/tunable/gemm_tunable.cuh [hunk 5]
```diff
@@ -78,6 +79,80 @@ class GemmTunableOp : public tunable::TunableOp<GemmParams<T>> {
...
+      this->ops_.emplace_back(std::move(op));
+    }
+  }
+
+  const StridedBatchedGemmParams<T>* PreTuning(const StridedBatchedGemmParams<T>* params) override {
+    if (!IsZero(params->beta)) {
+      // See GemmTunableOp<T>::PreTuning for more details
+      StridedBatchedGemmParams<T>* proxy = new StridedBatchedGemmParams<T>();
+      *proxy = *params;
+      HIP_CALL_THROW(hipMalloc(&(proxy->c), proxy->m * proxy->stride_c * sizeof(T)));
```
Comment (abudup) [inline onnxruntime/core/providers/rocm/tunable/gemm_tunable.cuh:141]: > `hipMalloc(&(proxy->c), proxy->m * proxy->stride_c * sizeof(T))` 

Shouldn't this be `hipMalloc(&(proxy->c), proxy->m * proxy->n * proxy->batch * sizeof(T))`. i.e., `proxy->c` should have size `m x n x batch_size` elems, whereas with your computation, `proxy->c` would have size ~`m x (m x n)` elems, because `proxy->stride_c = m x n x sizeof(T) + padding`, I'd assume?

File: onnxruntime/core/providers/rocm/tunable/gemm_tunable.cuh [hunk 6]
```diff
@@ -78,6 +79,80 @@ class GemmTunableOp : public tunable::TunableOp<GemmParams<T>> {
...
+      return proxy;
```
Comment (abudup) [inline onnxruntime/core/providers/rocm/tunable/gemm_tunable.cuh:142]: Also, consider copying the _data_ in `params->c` to `proxy->c` as well.

File: onnxruntime/core/providers/rocm/tunable/gemm.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (abudup) [reply onnxruntime/core/providers/rocm/tunable/gemm.cu:96]: I think there was a miscommunication here: I was suggesting that the code that I proposed (in my first comment in this thread) wouldn't compile. I wasn't talking about the original code. 

The intention of my original comment was to eliminate this `if () ... else` block that occurs in multiple places with a one-line `params.beta = normalize_scalar<T, ScalarT>(beta)`. I understand that the compiler will optimize away the `if()... else` block in your implementation. I was aiming for simpler (from a human's POV) code :-)

File: onnxruntime/python/tools/kernel_explorer/kernels/batched_gemm_test.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+):
+    assert dtype in ["float32", "float16"]
+
+    a_shape = (k, m) if transa else (m, k)
+    b_shape = (n, k) if transb else (k, n)
+
+    np.random.seed(0)
+    as_ = [(np.random.rand(*a_shape) + 0.5).astype(dtype).astype("float64") for i in range(batch)]
+    bs = [(np.random.rand(*b_shape) + 0.5).astype(dtype).astype("float64") for i in range(batch)]
+    intermidiate_cs = [(as_[i].T if transa else as_[i]) @ (bs[i].T if transb else bs[i]) for i in range(batch)]
```
Comment (abudup) [inline onnxruntime/python/tools/kernel_explorer/kernels/batched_gemm_test.py:34]: > `intermidiate_cs`

Typo: `intermidiate_cs` -> `intermediate_cs`

File: onnxruntime/python/tools/kernel_explorer/kernels/gemm_test.py [hunk 8]
```diff
@@ -29,14 +29,18 @@ def _test_gemm(func, dtype: str, m: int, n: int, k: int, transa=False, transb=Fa
...
+    else:
+        ref_c = alpha * intermidiate_c + beta * np.ones_like(intermidiate_c)
 
     bound = get_gemm_bound(dtype, a, b, ref_c, transa, transb)
 
     a = a.astype(dtype)
     b = b.astype(dtype)
 
-    my_c = np.zeros((m, n), dtype=dtype)
+    my_c = np.ones((m, n), dtype=dtype)
```
Comment (abudup) [inline onnxruntime/python/tools/kernel_explorer/kernels/gemm_test.py:43]: Can you please elaborate why this change is needed?

File: onnxruntime/core/providers/rocm/tunable/gemm_tunable.cuh [hunk 9]
```diff
@@ -78,6 +79,80 @@ class GemmTunableOp : public tunable::TunableOp<GemmParams<T>> {
...
 
+template <typename T, typename ALayout, typename BLayout>
+class BatchedGemmTunableOp : public tunable::TunableOp<BatchedGemmParams<T>> {
+ public:
+  BatchedGemmTunableOp() {
+    this->ops_.emplace_back(RocBlasBatchedGemmOp<T>);
+  }
+
+  const BatchedGemmParams<T>* PreTuning(const BatchedGemmParams<T>* params) override {
+    if (!IsZero(params->beta)) {
```
[the remark this answers is not in the document]
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/tunable/gemm_tunable.cuh:90]: It is uncessary because we iterate over it many times.

File: onnxruntime/core/providers/rocm/tunable/gemm_tunable.cuh (hunk 5 above)
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/tunable/gemm_tunable.cuh:141]: It is bugged here, this should be `batch * stride_c`.

But no, you must not use `proxy->m * proxy->n * proxy->batch` to compute. When the user call it with a stride, they mean it, so the padding is necessary, otherwise, you offset into an unknown wasted land.

File: onnxruntime/python/tools/kernel_explorer/kernels/gemm_test.py (hunk 8 above)
Comment (cloudhan) [reply onnxruntime/python/tools/kernel_explorer/kernels/gemm_test.py:43]: To test beta

File: onnxruntime/core/providers/rocm/tunable/gemm_tunable.cuh (hunk 6 above)
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/tunable/gemm_tunable.cuh:142]: Not considered at the moment =)

File: onnxruntime/core/providers/rocm/math/einsum_utils/einsum_auxiliary_ops.cc [hunk 10]
```diff
@@ -2,6 +2,7 @@
 // Licensed under the MIT License.
 
 #include "core/providers/shared_library/provider_api.h"
+#include "core/providers/rocm/tunable/gemm.h"
```
Comment (PeixuanZuo) [inline onnxruntime/core/providers/rocm/math/einsum_utils/einsum_auxiliary_ops.cc:5]: Could you please order these two header files?

File: onnxruntime/core/providers/rocm/tunable/gemm.cu [hunk 11]
```diff
@@ -108,6 +260,22 @@ GEMM(BFloat16, BFloat16) { return CALL_GEMM_WITH_AB_SWAPPED(BFloat16, BFloat16);
...
+STRIDED_BATCHED_GEMM(double,   double  ) { return CALL_STRIDED_BATCHED_GEMM_WITH_AB_SWAPPED(double,   double  ); }
+STRIDED_BATCHED_GEMM(float,    float   ) { return CALL_STRIDED_BATCHED_GEMM_WITH_AB_SWAPPED(float,    float   ); }
+STRIDED_BATCHED_GEMM(half,     half    ) { return CALL_STRIDED_BATCHED_GEMM_WITH_AB_SWAPPED(half,     half    ); }
+STRIDED_BATCHED_GEMM(BFloat16, BFloat16) { return CALL_STRIDED_BATCHED_GEMM_WITH_AB_SWAPPED(BFloat16, BFloat16); }
+STRIDED_BATCHED_GEMM(double,   float   ) { return CALL_STRIDED_BATCHED_GEMM_WITH_AB_SWAPPED(double,   float   ); }
+STRIDED_BATCHED_GEMM(half,     float   ) { return CALL_STRIDED_BATCHED_GEMM_WITH_AB_SWAPPED(half,     float   ); }
+STRIDED_BATCHED_GEMM(BFloat16, float   ) { return CALL_STRIDED_BATCHED_GEMM_WITH_AB_SWAPPED(BFloat16, float   ); }
 // clang-format on
 
 #undef CALL_GEMM_WITH_AB_SWAPPED
```
Comment (PeixuanZuo) [inline onnxruntime/core/providers/rocm/tunable/gemm.cu:281]: Do we need `#undef CALL_BATCHED_GEMM_WITH_AB_SWAPPED` and `#undef CALL_STRIDED_BATCHED_GEMM_WITH_AB_SWAPPED` here?

File: onnxruntime/python/tools/kernel_explorer/kernels/batched_gemm_test.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+    func, dtype: str, transa: bool, transb: bool, m: int, n: int, k: int, batch: int, alpha=1.0, beta=0.0
+):
+    assert dtype in ["float32", "float16"]
+
+    a_shape = (k, m) if transa else (m, k)
+    b_shape = (n, k) if transb else (k, n)
+
+    np.random.seed(0)
+    as_ = [(np.random.rand(*a_shape) + 0.5).astype(dtype).astype("float64") for i in range(batch)]
+    bs = [(np.random.rand(*b_shape) + 0.5).astype(dtype).astype("float64") for i in range(batch)]
```
Comment (PeixuanZuo) [inline onnxruntime/python/tools/kernel_explorer/kernels/batched_gemm_test.py:33]: why they named as `as_` and `bs` instead of `as_` and `bs_`?

File: onnxruntime/core/providers/rocm/math/einsum_utils/einsum_auxiliary_ops.cc (hunk 10 above)
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/math/einsum_utils/einsum_auxiliary_ops.cc:5]: No.

File: onnxruntime/python/tools/kernel_explorer/kernels/batched_gemm_test.py (hunk 12 above)
Comment (cloudhan) [reply onnxruntime/python/tools/kernel_explorer/kernels/batched_gemm_test.py:33]: `as` is a keyword in python...

File: onnxruntime/core/providers/rocm/tunable/gemm_tunable.cuh (hunk 5 above)
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/tunable/gemm_tunable.cuh]: Fixed.

File: onnxruntime/python/tools/kernel_explorer/kernels/batched_gemm_test.py (hunk 7 above)
Comment (cloudhan) [reply onnxruntime/python/tools/kernel_explorer/kernels/batched_gemm_test.py]: Fixed

**Comment by cloudhan** [comment]
FYI, ck's `GroupedGemm`s are batched gemms in rocblas and ck's `BatchedGemm`s are strided batched gemms in rocblas. And for ck's `GroupedGemm` it requires host array of device pointers and rocblas requires device array of device pointers. We don't want to fix the API disparity on our side at the moment. So GroupedGemm` is not included at the moment.

**Review by abudup — DISMISSED** [review]

**Review by abudup — DISMISSED** [review]
Looks good, thanks!

**Review by PeixuanZuo — APPROVED** [review]
